Add esTypes property to index pattern field

Bargs opened the pull request:
## Summary

Adds an array property to the Field object containing all of the types the field is mapped as across ES indices. 

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

~~- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)~~
~~- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)~~
~~- [ ] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials~~
- [x] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
~~- [ ] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)~~

### For maintainers

~~- [ ] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)~~
~~- [ ] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)~~

elasticmachine commented:
Pinging @elastic/kibana-app

wylieconlon commented:
This approach makes sense to me. It was surprising to me that this information was missing when working on the new editor. I assume this needs some kind of testing or migration plan to be ready, but I'm not super familiar with the requirements there.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10118/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10343/)

Bargs commented:
Fixed and added tests and updated unit test fixtures and sample data. 

However I did not touch the functional test fixtures. These can't be easily updated automatically afaik. We'd need to load up each data set, refresh the index pattern and save it to the archive in order to get accurate type information from the field_caps API. That's a lot of up front work when none of the current functional tests require this information, and they could be updated one by one as needed. This seems to be the approach taken when similar updates have been made, for example `readFromDocValues` is missing from some of the older test fixtures like discover.

Unless reviewers strongly disagree with the reasoning above, I'm pretty much done with the code once CI passes and this is ready for review.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10455/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10464/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10475/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10582/)

Bargs commented:
I thought about that back when I was adding multi field info to the index pattern. [Nathan thought](https://github.com/elastic/kibana/pull/33681#issuecomment-479575319) it would be a bad idea to automatically refresh index patterns and I agreed. At least for now I think we should build features with the assumption that this info might not be available. The typescript definition marks this property as optional so as long as someone is using TS they'll be forced to handle the situation where this property is undefined. I'll add a comment to the TS definition explaining why it could be undefined.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10640/)
